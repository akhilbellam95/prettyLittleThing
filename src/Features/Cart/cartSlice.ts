import { createSlice } from '@reduxjs/toolkit';

interface CartItem {
  id: number;
  quantity: number;
};

export interface CartState {
  cartItems: CartItem[],
  total: number
}

export const cartInitialState: CartState = {
  cartItems: [] as CartItem[],
  total: 0,
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState: cartInitialState,
  reducers: {
    addToCart: (state, action) => {
      const { id } = action.payload;
      const existingItem = state.cartItems.find((item) => item.id === id);
      let total = state.total;
      if (existingItem) {
        existingItem.quantity++;
      } else {
        state.cartItems.push({ id, quantity: 1 });
      }
      total += action.payload.price;
      state.total = parseFloat(total.toFixed(2));
    },
    removeFromCart: (state, action) => {
      const itemIndex = state.cartItems.findIndex((el) => el.id === action.payload.id);
      if (itemIndex !== -1) {
        const item = state.cartItems[itemIndex];
        let total = state.total;
        if (item.quantity === 1) {
          state.cartItems.splice(itemIndex, 1);
        } else {
          item.quantity--;
        }
        total -= action.payload.price;
        state.total = parseFloat(total.toFixed(2));
      }
    },
    updateTotal: (state, action) => {
      state.total = action.payload;
    },
  },
});

export const { addToCart, removeFromCart, updateTotal } = cartSlice.actions;

export default cartSlice.reducer;
