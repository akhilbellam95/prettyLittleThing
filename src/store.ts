// eslint-disable-next-line prettier/prettier
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './Features/Cart/cartSlice';
import productReducer from './Features/Product/productSlice';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    product: productReducer,
  },
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;

export type RootState = ReturnType<typeof store.getState>;
