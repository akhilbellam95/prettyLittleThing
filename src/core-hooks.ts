import { configureStore, createStore } from "@reduxjs/toolkit";
import { useSelector } from "react-redux";
import { TypedUseSelectorHook, useDispatch } from "react-redux";
import { CartState } from "./Features/Cart/cartSlice";
import { ProductsState } from "./Features/Product/productSlice";

export interface ApplicationState {
  cart: CartState;
  product: ProductsState;
}

export const useAppDispatch = () =>
  useDispatch<ReturnType<typeof configureStore>['dispatch']>();

export const useAppSelector: TypedUseSelectorHook<ApplicationState> =
  useSelector;