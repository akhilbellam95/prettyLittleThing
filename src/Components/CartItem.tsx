import React from 'react';
import { Image, Text, View } from 'react-native';
import { addToCart, removeFromCart } from '../Features/Cart/cartSlice';
import AddRemoveItem from './AddRemoveItem';
import { ApplicationState, useAppDispatch, useAppSelector } from '../core-hooks';
import { ItemProps, styles } from './CartItems';

export const CartItem = ({ cartItem }: ItemProps) => {
  const { products } = useAppSelector((state: ApplicationState) => state.product) || [];

  const dispatch = useAppDispatch();

  const handleAddToCart = (product) => {
    dispatch(addToCart({ ...product, id: cartItem.id, quantity: 1 }));
  };

  const handleRemoveFromCart = (product) => {
    dispatch(removeFromCart({ ...product, id: cartItem.id, quantity: 1 }));
  };

  return (
    <View>
      {products?.length > 0 && (
        <>
          {products.map(
            (product: {
              id: number;
              img: string;
              name: string;
              colour: string;
              price: number;
            }): JSX.Element | null =>
              cartItem.id === product.id ? (
                <View style={styles.cartItem}>
                  <Image
                    source={{ uri: product.img }}
                    style={{ height: 180, width: 120, marginRight: 20 }}
                  />
                  <View style={{ justifyContent: 'space-between' }}>
                    <View>
                      <Text style={{ maxWidth: 160 }}>{'Name: ' + product.name}</Text>
                      <Text style={{ maxWidth: 160, marginTop: 5, marginBottom: 5 }}>
                        {'Colour: ' + product.colour}
                      </Text>
                      <Text style={{ maxWidth: 160 }}>{'Price: ' + product.price}</Text>
                    </View>

                    <AddRemoveItem
                      style={{ marginTop: 10 }}
                      addPressed={() => handleAddToCart(product)}
                      removePressed={() => handleRemoveFromCart(product)}
                      quantity={cartItem.quantity}
                    />
                  </View>
                </View>
              ) : null,
          )}
        </>
      )}
    </View>
  );
};
