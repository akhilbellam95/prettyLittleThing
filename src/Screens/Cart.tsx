import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import CartItems from '../Components/CartItems';
import { ApplicationState, useAppSelector } from '../core-hooks';

const Cart = () => {
  const { cartItems } = useAppSelector((state: ApplicationState) => state.cart);

  return (
    <View style={styles.container}>
      {cartItems.length > 0 ? <CartItems /> : <Text>Cart is empty</Text>}
    </View>
  );
};

export default Cart;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
