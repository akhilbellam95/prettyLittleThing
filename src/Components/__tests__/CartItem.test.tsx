import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '../../store';
import { useAppDispatch, useAppSelector } from '../../core-hooks';
import { CartItem } from '../CartItem';
import { addToCart } from '../../Features/Cart/cartSlice';
import { mockProducts } from '../../mocks/data';

jest.mock('../../core-hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: jest.fn(),
}));

describe('CartItem component testing', () => {
  const mockDispatch = jest.fn();
  const mockSelector = jest.fn();

  beforeAll(() => {
    (useAppSelector as jest.Mock).mockImplementation(mockSelector);
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  const snap: unknown = render(
    <Provider store={store}>
      <CartItem cartItem={{ id: 1, quantity: 1 }} />
    </Provider>,
  ).toJSON();

  beforeEach(() => {
    mockSelector.mockReturnValue({
      products: mockProducts,
    });
  });

  it('renders correctly', () => {
    expect(snap).toMatchSnapshot();
  });

  it('should add product to cart', () => {
    const product = { id: 1 };
    store.dispatch(addToCart(product));
    const state = store.getState()['cart'];
    expect(state.cartItems).toHaveLength(1);
    expect(state.cartItems[0]).toEqual({ ...product, quantity: 1 });

    expect(snap).toMatchSnapshot();
  });

  it('should call when user press add button', () => {
    const { getByTestId } = render(<CartItem cartItem={{ id: 1, quantity: 1 }} />);
    const addButton = getByTestId('add-btn');

    fireEvent.press(addButton);
  });

  it('should call when user press remove button', () => {
    const { getByTestId } = render(<CartItem cartItem={{ id: 1, quantity: 1 }} />);
    const removeButton = getByTestId('remove-btn');

    fireEvent.press(removeButton);
  });

  // it('should remove product from cart', () => {
  //   const product = { id: 1 };
  //   store.dispatch(addToCart(product));
  //   store.dispatch(removeFromCart(product));
  //   const state = store.getState()['cart'];
  //   expect(state.cartItems).toHaveLength(0);
  // });
});
