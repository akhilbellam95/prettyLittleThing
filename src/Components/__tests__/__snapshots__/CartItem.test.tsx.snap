// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component testing renders correctly 1`] = `<View />`;

exports[`CartItem component testing should add product to cart 1`] = `<View />`;
